src/Lecture2.hs

```diff
@@ -207,8 +440,15 @@ The algorithm of merge sort is the following:
 >>> mergeSort [3, 1, 2]
 [1,2,3]
 -}
-mergeSort :: [Int] -> [Int]
-mergeSort = error "TODO"
+-- Based on @msort@ as implemented in Listing 21.5 of Programming in Scala, Fifth Edition, by Odersky et al.
+mergeSort :: Ord a => [a] -> [a]
+mergeSort [] = []
+mergeSort (x : []) = [x]
+mergeSort xs =
+    merge (mergeSort as) (mergeSort bs)
+    where
+        n = div (length xs) 2
+        (as, bs) = splitAt n xs
```

Comment:
You can avoid using `length` if you notice that you don't necessarily need to split in the middle. You only need to split it into two lists of approximately the same size.

In that case, you can implement a solution that traverses a list only once and splits its elements into two lists of elements on only even positions and only odd positions.

Reply:
does Haskell have something like `groupBy`, that is, a function taking a list and a predicate and returning a pair of lists, one with the passing elements and one with the non-passing ones?

Reply:
Yes, Haskell does. Such a function is called `partition`:

* https://hackage.haskell.org/package/base-4.16.0.0/docs/Data-List.html#v:partition